content/blog/testing-kafka-connectors.asciidoc

```diff
@@ -0,0 +1,317 @@
+Naturally, this raises the question of how to test such a Kafka connector,
+making sure it propagates the data between the connected external system and Kafka correctly and completely.
+In this blog post I'd like to focus on testing approaches for Kafka Connect _source connectors_,
+i.e. connectors like https://debezium.io/[Debezium], which ingest data from an external system into Kafka.
+
+<!--more-->
+
+== Unit Tests
+
+One first obvious approach is implementing good old unit tests:
+simply instantiate the class under test (typically, your `SourceConnector` or `SourceTask` implementation),
+invoke its methods (e.g. `SourceConnector::taskConfigs()` or `SourceTask::poll()`),
+and assert the return values.
+
+Here's an example for such a test from https://github.com/gunnarmorling/kcetcd[kc-etcd], a simple source connector for https://etcd.io/[Etcd].
+Etcd is a distributed key/value store,
+for instance used by Kubernetes as its metadata storage.
+Note that kc-etcd isn't meant to be a production-ready connector;
+I have written it primarily for learning and teaching purposes.
+
+This test verifies that the connector produces the correct task configuration,
+based on the given number of maximum tasks:
+
+[source,java,linenums=true]
+----
+public class EtcdSourceConnectorTest {
+
+  @Test
+  public void shouldCreateConfigurationForTasks() throws Exception {
+    EtcdSourceConnector connector = new EtcdSourceConnector();
+    Map<String, String> config = new HashMap<>();
+    config.put(
+      "clusters", 
+      "etcd-a=http://etcd-a-1:2379,http://etcd-a-2:2379,http://etcd-a-3:2379;etcd-b=http://etcd-b-1:2379;etcd-c=http://etcd-c-1:2379"
+    ); // <1>
+    connector.start(config);
+
+    List<Map<String, String>> taskConfigs = connector.taskConfigs(2); // <2>
+    assertThat(taskConfigs).hasSize(2);
+
+    // <3>
+    taskConfig = taskConfigs.get(0);
+    assertThat(taskConfig).containsEntry("clusters",
+        "etcd-a=http://etcd-a-1:2379,http://etcd-a-2:2379,http://etcd-a-3:2379;etcd-b=http://etcd-b-1:2379");
+
+    // <4>
+    taskConfig = taskConfigs.get(1);
+    assertThat(taskConfig).containsEntry("clusters",
+        "etcd-c=http://etcd-c-1:2379");
+  }
+}
+----
+<1> Configure the connector with three Etcd clusters
+<2> Request the configuration for two tasks
+<3> The first task should handle the first two clusters
+<4> The second task should handle the remaining third cluster
+
+Things looks similar for testing the actual polling loop of the connector's task class.
+As this is about testing a source connector,
+we first need to do some data changes in the configured Etcd cluster(s),
+before we can assert the corresponding ``SourceRecord``s emitted by the task.
+As part of kc-etcd, I've implemented a _very_ basic testing harness named https://github.com/gunnarmorling/kcetcd/tree/main/src/test/java/dev/morling/kcute[kcute]
+("Kafka Connect Unit Testing") which simplifies that task a bit.
```

Comment:
Maybe replace the word "task" with "process" since the word task comes up a lot already when talking about connect.

Reply:
LOL, yeah, good point. Changing it.